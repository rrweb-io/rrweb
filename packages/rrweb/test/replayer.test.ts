/* tslint:disable no-string-literal no-console */

import * as fs from 'fs';
import * as path from 'path';
import * as puppeteer from 'puppeteer';
import {
  assertDomSnapshot,
  launchPuppeteer,
  sampleEvents as events,
  sampleStyleSheetRemoveEvents as stylesheetRemoveEvents,
} from './utils';
import styleSheetRuleEvents from './events/style-sheet-rule-events';
import orderingEvents from './events/ordering';

interface ISuite {
  code: string;
  browser: puppeteer.Browser;
  page: puppeteer.Page;
}

describe('replayer', function () {
  jest.setTimeout(10_000);

  let code: ISuite['code'];
  let browser: ISuite['browser'];
  let page: ISuite['page'];

  beforeAll(async () => {
    browser = await launchPuppeteer();

    const bundlePath = path.resolve(__dirname, '../dist/rrweb.min.js');
    code = fs.readFileSync(bundlePath, 'utf8');
  });

  beforeEach(async () => {
    page = await browser.newPage();
    await page.goto('about:blank');
    await page.evaluate(code);
    await page.evaluate(`let events = ${JSON.stringify(events)}`);

    page.on('console', (msg) => console.log('PAGE LOG:', msg.text()));
  });

  afterEach(async () => {
    await page.close();
  });

  afterAll(async () => {
    await browser.close();
  });

  it('can get meta data', async () => {
    const meta = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.getMetaData();
    `);
    expect(meta).toEqual({
      startTime: events[0].timestamp,
      endTime: events[events.length - 1].timestamp,
      totalTime: events[events.length - 1].timestamp - events[0].timestamp,
    });
  });

  it('will start actions when play', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play();
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(events.length);
  });

  it('will clean actions when pause', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play();
      replayer.pause();
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(0);
  });

  it('can play at any time offset', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play(1500);
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(
      events.filter((e) => e.timestamp - events[0].timestamp >= 1500).length,
    );
  });

  it('can play a second time in the future', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play(500);
      replayer.play(1500);
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(
      events.filter((e) => e.timestamp - events[0].timestamp >= 1500).length,
    );
  });

  it('can play a second time to the past', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play(1500);
      replayer.play(500);
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(
      events.filter((e) => e.timestamp - events[0].timestamp >= 500).length,
    );
  });

  it('can pause at any time offset', async () => {
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.pause(2500);
      replayer['timer']['actions'].length;
    `);
    const currentTime = await page.evaluate(`
      replayer.getCurrentTime();
    `);
    const currentState = await page.evaluate(`
      replayer['service']['state']['value'];
    `);
    expect(actionLength).toEqual(0);
    expect(currentTime).toEqual(2500);
    expect(currentState).toEqual('paused');
  });

  it('can fast forward past StyleSheetRule changes on virtual elements', async () => {
    await page.evaluate(`events = ${JSON.stringify(styleSheetRuleEvents)}`);
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play(1500);
      replayer['timer']['actions'].length;
    `);

    expect(actionLength).toEqual(
      styleSheetRuleEvents.filter(
        (e) => e.timestamp - styleSheetRuleEvents[0].timestamp >= 1500,
      ).length,
    );

    await assertDomSnapshot(
      page,
      __filename,
      'style-sheet-rule-events-play-at-1500',
    );
  });

  it('should apply fast forwarded StyleSheetRules that where added', async () => {
    await page.evaluate(`events = ${JSON.stringify(styleSheetRuleEvents)}`);
    const result = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.pause(1500);
      const rules = [...replayer.iframe.contentDocument.styleSheets].map(
        (sheet) => [...sheet.rules],
      ).flat();
      rules.some((x) => x.selectorText === '.css-added-at-1000-deleted-at-2500');
    `);

    expect(result).toEqual(true);
  });

  it('can handle removing style elements', async () => {
    await page.evaluate(`events = ${JSON.stringify(stylesheetRemoveEvents)}`);
    const actionLength = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play(2500);
      replayer['timer']['actions'].length;
    `);
    expect(actionLength).toEqual(
      stylesheetRemoveEvents.filter(
        (e) => e.timestamp - stylesheetRemoveEvents[0].timestamp >= 2500,
      ).length,
    );

    await assertDomSnapshot(
      page,
      __filename,
      'style-sheet-remove-events-play-at-2500',
    );
  });

  it('can fast forward past StyleSheetRule deletion on virtual elements', async () => {
    await page.evaluate(`events = ${JSON.stringify(styleSheetRuleEvents)}`);

    await assertDomSnapshot(
      page,
      __filename,
      'style-sheet-rule-events-play-at-2500',
    );
  });

  it('should delete fast forwarded StyleSheetRules that where removed', async () => {
    await page.evaluate(`events = ${JSON.stringify(styleSheetRuleEvents)}`);
    const result = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.pause(3000);
      const rules = [...replayer.iframe.contentDocument.styleSheets].map(
        (sheet) => [...sheet.rules],
      ).flat();
      rules.some((x) => x.selectorText === '.css-added-at-1000-deleted-at-2500');
    `);

    expect(result).toEqual(false);
  });

  it('can stream events in live mode', async () => {
    const status = await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events, {
        liveMode: true
      });
      replayer.startLive();
      replayer.service.state.value;
    `);
    expect(status).toEqual('live');
  });

  it('replays same timestamp events in correct order', async () => {
    await page.evaluate(`events = ${JSON.stringify(orderingEvents)}`);
    await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events);
      replayer.play();
    `);
    await page.waitForTimeout(50);

    await assertDomSnapshot(page, __filename, 'ordering-events');
  });

  it('replays same timestamp events in correct order (with addAction)', async () => {
    await page.evaluate(`events = ${JSON.stringify(orderingEvents)}`);
    await page.evaluate(`
      const { Replayer } = rrweb;
      const replayer = new Replayer(events.slice(0, events.length-2));
      replayer.play();
      replayer.addEvent(events[events.length-2]);
      replayer.addEvent(events[events.length-1]);
    `);
    await page.waitForTimeout(50);

    await assertDomSnapshot(page, __filename, 'ordering-events');
  });
});
