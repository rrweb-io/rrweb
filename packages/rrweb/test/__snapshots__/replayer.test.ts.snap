// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replayer can fast forward past StyleSheetRule changes on virtual elements 1`] = `
"file-frame-4
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  </head>
  <body>
    <div class=\\"replayer-wrapper\\">
      <div class=\\"replayer-mouse\\"></div>
      <canvas
        class=\\"replayer-mouse-tail\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit\\"
      ></canvas
      ><iframe
        sandbox=\\"allow-same-origin\\"
        scrolling=\\"no\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit; pointer-events: none\\"
      ></iframe>
    </div>
  </body>
</html>


file-frame-5
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-0\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-1\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-2\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-3\\" />
  </head>
  <body>
    <a class=\\"css-added-at-1000-deleted-at-2500\\">string</a>
  </body>
</html>


file-cid-0
@charset \\"utf-8\\";

.rr-block { background: currentcolor; }

noscript { display: none !important; }

html.rrweb-paused *, html.rrweb-paused ::before, html.rrweb-paused ::after { animation-play-state: paused !important; }


file-cid-1
@charset \\"utf-8\\";

.css-added-at-400 { padding: 1.3125rem; flex: 0 0 auto; width: 100%; }


file-cid-2
@charset \\"utf-8\\";

.css-added-at-200-overwritten-at-3000 { opacity: 1; transform: translateX(0px); }

.css-added-at-500-overwritten-at-3000 { border: 1px solid blue; }


file-cid-3
@charset \\"utf-8\\";

.css-added-at-200 { position: fixed; top: 0px; right: 0px; left: 4rem; z-index: 15; flex-shrink: 0; height: 0.25rem; overflow: hidden; background-color: rgb(17, 171, 209); }

.css-added-at-200.alt { height: 0.25rem; background-color: rgb(190, 232, 242); opacity: 0; transition: opacity 0.5s ease 0s; }

.css-added-at-1000-deleted-at-2500 { display: flex; flex-direction: column; min-width: 60rem; min-height: 100vh; color: blue; }

.css-added-at-200.alt2 { padding-left: 4rem; }
"
`;

exports[`replayer can fast forward past StyleSheetRule deletion on virtual elements 1`] = `
"file-frame-0
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  </head>
  <body></body>
</html>
"
`;

exports[`replayer can handle removing style elements 1`] = `
"file-frame-1
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  </head>
  <body>
    <div class=\\"replayer-wrapper\\">
      <div class=\\"replayer-mouse\\"></div>
      <canvas
        class=\\"replayer-mouse-tail\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit\\"
      ></canvas
      ><iframe
        sandbox=\\"allow-same-origin\\"
        scrolling=\\"no\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit; pointer-events: none\\"
      ></iframe>
    </div>
  </body>
</html>


file-frame-2
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-0\\" />
  </head>
  <body></body>
</html>


file-cid-0
@charset \\"utf-8\\";

.rr-block { background: currentcolor; }

noscript { display: none !important; }

html.rrweb-paused *, html.rrweb-paused ::before, html.rrweb-paused ::after { animation-play-state: paused !important; }
"
`;

exports[`replayer replays same timestamp events in correct order (with addAction) 1`] = `
"file-frame-1
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  </head>
  <body>
    <div class=\\"replayer-wrapper\\">
      <div class=\\"replayer-mouse\\"></div>
      <canvas
        class=\\"replayer-mouse-tail\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit\\"
      ></canvas
      ><iframe
        sandbox=\\"allow-same-origin\\"
        scrolling=\\"no\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit; pointer-events: none\\"
      ></iframe>
    </div>
  </body>
</html>


file-frame-2
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-0\\" />
  </head>
  <body>
    <span>Final - correct</span>
  </body>
</html>


file-cid-0
@charset \\"utf-8\\";

.rr-block { background: currentcolor; }

noscript { display: none !important; }

html.rrweb-paused *, html.rrweb-paused ::before, html.rrweb-paused ::after { animation-play-state: paused !important; }
"
`;

exports[`replayer replays same timestamp events in correct order 1`] = `
"file-frame-1
<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  </head>
  <body>
    <div class=\\"replayer-wrapper\\">
      <div class=\\"replayer-mouse\\"></div>
      <canvas
        class=\\"replayer-mouse-tail\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit\\"
      ></canvas
      ><iframe
        sandbox=\\"allow-same-origin\\"
        scrolling=\\"no\\"
        width=\\"1000\\"
        height=\\"800\\"
        style=\\"display: inherit; pointer-events: none\\"
      ></iframe>
    </div>
  </body>
</html>


file-frame-2
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"file-cid-0\\" />
  </head>
  <body>
    <span>Final - correct</span>
  </body>
</html>


file-cid-0
@charset \\"utf-8\\";

.rr-block { background: currentcolor; }

noscript { display: none !important; }

html.rrweb-paused *, html.rrweb-paused ::before, html.rrweb-paused ::after { animation-play-state: paused !important; }
"
`;
